compute_endpoint/globus_compute_endpoint/self_diagnostic.py

```diff
@@ -19,26 +18,31 @@
 from globus_compute_sdk.sdk.web_client import WebClient
 
 
-def cat(path: str, wildcard: bool = False, max_bytes: int | None = None):
-    def kernel() -> None:
-        full_path = os.path.expanduser(path)
+def cat(path: str, wildcard: bool = False, max_bytes: int = 0):
+    max_bytes = max(0, max_bytes)
 
-        if wildcard:
-            files = glob.glob(full_path)
-        else:
-            files = [full_path]
+    full_path = os.path.expanduser(path)
+    if wildcard:
+        files = glob.glob(full_path)
+    else:
+        files = [full_path]
 
+    def kernel() -> None:
         for filename in files:
-            click.echo("cat " + filename)
+            cat_cmd = "cat " + filename
+            hline = "=" * len(cat_cmd)
+            click.echo(cat_cmd + "\n" + hline)
             if os.path.exists(filename):
                 with open(filename, "rb") as f:
                     if max_bytes:
                         file_size = os.fstat(f.fileno()).st_size
                         f.seek(max(file_size - max_bytes, 0))
-                    content = f.read().decode("utf-8")
-                click.echo(textwrap.indent(content, "  "))
+                    content = f.read().replace(b"\n", b"\n | ")
```

Comment:
macOS uses `\r` as its newline format. I don't know if that's relevant in this context but wanted to mention it.

Also, I see that this gets output to the terminal but it's left to click to deal with encodings. I recommend testing against Unicode content that's been cut off mid-encoding by `max_bytes`. For example:

```python
"\N{winking face}".encode("utf-8")[1:].decode("utf-8")  # UnicodeDecodeError
```

Reply:
> macOS uses `\r` as its newline format. I don't know if that's relevant in this context but wanted to mention it.

Is that still true?  I know that was the case when we were growing up, but I've not had to deal with that in C++ land in a very, _very_ long time.  (Where by "C++", I mean my personal experience handling raw strings in log files, unfettered by Python's shims.)  I don't recall having to deal with that in almost 20 years.

I'll poke at that in the morning (and maybe ask a macOS user on our team to verify!).  Thanks for bringing it up.

> Also, I see that this gets output to the terminal but it's left to click to deal with encodings. I recommend testing against Unicode content that's been cut off mid-encoding by max_bytes. For example:

This is a much stronger point (and kudos for it!), but I'm not concerned about it at this juncture given these three points:

1. The use-case for this code is in collecting diagnostic information from a user's setup without us having to ask for all the things.  This is mostly to get context and low-hanging fruit visuals on the end-users setup.
2. Unicode points are isolated.  Meaning that one malformed character will not prevent the following character from being interpreted correctly.  For reference, I point out [`::from _utf8_lossy()`](https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8_lossy) in Rust-land, and the second argument to Python's [`bytes.decode()`](https://docs.python.org/3/library/stdtypes.html#bytes.decode).
3. This script collects, by default, up to 5MiB of a log file&nbsp;&mdash;&nbsp;well beyond most errors we currently see.

> "\N{winking face}"

Okay, that's a neat little tidbit I didn't know about.  Very cool.

Reply:
I think we may be talking past each other re: Unicode. However, I've confirmed that `click.echo()` decodes the byte strings using an `errors="replace"` strategy, so if the first byte of the sequence were lopped off by the shifting `max_bytes` window it won't cause a crash:

```pycon
>>> click.echo("\N{winking face}abc".encode("utf-8")[1:])
���abc
```

My concern regarding a Unicode bytes sequence getting cut off (and throwing a `UnicodeDecodeError`) is satisfied.

Regarding `\r` being used on macOS...it isn't as of macOS X. We tread on shifting sands...

Reply:
Ah ... I see your concern.  _Crashing_!  Heh.  Okay, how about random data rather than `b"a" * N`?  See the most recent addition.  I think that should assuage your concern / observation?

Reply:
Perfect! On my machine I'm getting 100% `UnicodeDecodeError` when running `os.urandom(100).decode('utf-8')`.

I imagine it'll be easy to detect an issue if click changes its behavior.